### PR TITLE
242 choose who to share location with

### PR DESCRIPTION
**Pull request by marinovdeyan**
Each user now has two extra arrays in their Firestore document that contain the people they share their location with (locationSharedWith) and the people who are sharing their location with them (locationSharedBy).

So far, you are able to:
- share location with someone which updates your locationSharedWith and the locationSharedBy of your friend
- stop sharing with someone which removes your friend from your locationSharedWith and removes yourself from locationSharedBy in your friend's document in Firestore
- get the locationSharedWith of someone
- get the locationSharedBy of someone

UserRepositoryFirestore and UserViewModel have functions to get or update these arrays.

Tests have also been written for the various functions. For the moment, no UI is accompanying these backend features.

### CODE REVIEW

File: app/src/test/java/com/swent/suddenbump/model/user/UserRepositoryFirestoreTest.kt [hunk 1]
```diff
@@ -2922,4 +2922,507 @@ class UserRepositoryFirestoreTest {
...
+    userRepositoryFirestore.getSharedWithFriends(
+        uid = uid,
+        onSuccess = { result ->
+          assertNotNull(result)
+          assertTrue(result.isEmpty())
+        },
+        onFailure = { fail("Should not fail when no data is available") })
+
+    shadowOf(Looper.getMainLooper()).idle()
+  }
```
Comment (DanR2003) [inline app/src/test/java/com/swent/suddenbump/model/user/UserRepositoryFirestoreTest.kt:3427]: Tests are very extensive, and also the comments makes them easy to read. Nice job!

**Review by DanR2003 — CHANGES_REQUESTED** [review]
- The functions written for sharing / stop sharing location with friends and for getting the locationSharedWith and the locationSharedBy of someone are very well implemented. 
- Also they are thoroughly tested and the comments added inside the tests makes following their steps much easier.
Overall, this is a qualitative PR in my opinion. The functions added will be very useful once we also add the UI for the functionalities they implement. After dealing with the requested changes (removing the residual files, and adding documentation),  this will be good to be merged into main.

File: app/.attach_pid92345
[the remark this answers is not in the document]
Comment (marinovdeyan) [reply app/.attach_pid92345:1]: Files removed.

File: app/src/main/java/com/swent/suddenbump/model/user/UserRepositoryFirestore.kt [hunk 2]
```diff
@@ -1079,6 +1079,146 @@ class UserRepositoryFirestore(private val db: FirebaseFirestore, private val con
...
+          if (result.data?.get("locationSharedWith") == null) {
+            emptyList<User>()
+          } else {
+            onSuccess(
+                documentSnapshotToUserList(
+                    result.data?.get("locationSharedWith").toString(), onFailure))
+          }
+        }
+  }
+
```
[the remark this answers is not in the document]
Comment (marinovdeyan) [reply app/src/main/java/com/swent/suddenbump/model/user/UserRepositoryFirestore.kt:1221]: Documentation added!

File: app/src/main/java/com/swent/suddenbump/model/user/UserViewModel.kt [hunk 3]
```diff
@@ -668,4 +666,32 @@ open class UserViewModel(
...
+    repository.stopSharingLocationWithFriend(uid, friend.uid, onSuccess, onFailure)
+  }
+
+  fun getLocationSharedBy(
+      uid: String,
+      onSuccess: (List<User>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    repository.getSharedByFriends(uid, onSuccess, onFailure)
+  }
```
[the remark this answers is not in the document]
Comment (marinovdeyan) [reply app/src/main/java/com/swent/suddenbump/model/user/UserViewModel.kt:696]: Documentation added!

**Comment by marinovdeyan** [comment]
@DanR2003 I have added the documentation and removed the residual files. Please review the code again and tell me if it's okay for you to merge :). Thanks!

**Review by DanR2003 — APPROVED** [review]
Perfect! Nice to see you did what I requested :)

**Pull request merged**